Userland/Services/SpiceAgent/SpiceAgent.cpp

```diff
@@ -1,310 +1,95 @@
-            bytes_copied += next_header.size;
-        }
-
-        m_just_set_clip = true;
-        if (type == ClipboardType::Text) {
-            if (data_buffer.is_empty()) {
-                m_clipboard_connection.async_set_clipboard_data({}, "text/plain", {});
-            } else {
-                auto anon_buffer = TRY(Core::AnonymousBuffer::create_with_size(data_buffer.size()));
-                memcpy(anon_buffer.data<void>(), data_buffer.data(), data_buffer.size());
-                m_clipboard_connection.async_set_clipboard_data(anon_buffer, "text/plain", {});
-            }
-            return {};
-        } else {
-            ErrorOr<Gfx::ImageFrameDescriptor> frame_or_error = Gfx::ImageFrameDescriptor {};
-            if (type == ClipboardType::PNG) {
-                if (Gfx::PNGImageDecoderPlugin::sniff({ data_buffer.data(), data_buffer.size() })) {
-                    auto png_decoder = TRY(Gfx::PNGImageDecoderPlugin::create({ data_buffer.data(), data_buffer.size() }));
-                    if (!png_decoder->initialize().is_error())
-                        frame_or_error = png_decoder->frame(0);
-                }
-            } else if (type == ClipboardType::BMP) {
-                if (Gfx::BMPImageDecoderPlugin::sniff({ data_buffer.data(), data_buffer.size() })) {
-                    auto bmp_decoder = TRY(Gfx::BMPImageDecoderPlugin::create({ data_buffer.data(), data_buffer.size() }));
-                    if (!bmp_decoder->initialize().is_error())
-                        frame_or_error = bmp_decoder->frame(0);
-                }
-            } else if (type == ClipboardType::JPEG) {
-                if (Gfx::JPEGImageDecoderPlugin::sniff({ data_buffer.data(), data_buffer.size() })) {
-                    auto jpeg_decoder = TRY(Gfx::JPEGImageDecoderPlugin::create({ data_buffer.data(), data_buffer.size() }));
-                    if (!jpeg_decoder->initialize().is_error())
-                        frame_or_error = jpeg_decoder->frame(0);
-                }
-            } else {
-                dbgln("Unknown clipboard type: {}", (u32)type);
-                return {};
-            }
-
-            auto const& bitmap = frame_or_error.value().image;
-            m_clipboard_connection.set_bitmap(*bitmap);
-        }
-        break;
-    }
     default:
-        dbgln("Unhandled message type {}", message->type);
+        dbgln("Unknown message received: {}", header);
+        break;
     }
 
     return {};
 }
 
-void SpiceAgent::read_n(void* dest, size_t n)
+ErrorOr<ByteBuffer> SpiceAgent::read_message_buffer()
 {
-    size_t bytes_read = 0;
-    while (bytes_read < n) {
-        int nread = read(m_fd, (u8*)dest + bytes_read, n - bytes_read);
-        if (nread > 0) {
-            bytes_read += nread;
-        } else if (errno == EAGAIN) {
-            continue;
-        } else {
-            dbgln("Failed to read: {}", errno);
-            return;
-        }
+    // TODO(Caoimhe): `ChunkHeader` class?
```

Comment:
Same comment about attribution on TODOs. This one could use some elaboration though